vladimir.ohantsev/task-5/pkg/conveyer/conveyer.go

```diff
@@ -0,0 +1,147 @@
+package conveyer
+
+import (
+	"context"
+	"errors"
+
+	"golang.org/x/sync/errgroup"
+)
+
+var ErrChanNotFound = errors.New("chan not found")
+
+type Pipeline struct {
+	size     int
+	channels map[string]chan string
+	handlers []func(ctx context.Context) error
+}
+
+func New(size int) *Pipeline {
+	if size < 0 {
+		panic("invalid chan size")
+	}
+
+	return &Pipeline{
+		size:     size,
+		channels: make(map[string]chan string),
+		handlers: []func(ctx context.Context) error{},
+	}
+}
+
+func (p *Pipeline) register(name string) chan string {
+	if ch, ok := p.channels[name]; ok {
+		return ch
+	}
+
+	ch := make(chan string, p.size)
+	p.channels[name] = ch
+
+	return ch
+}
+
+func (p *Pipeline) RegisterDecorator(
+	callback func(
+		ctx context.Context,
+		input chan string,
+		output chan string,
+	) error,
+	input string,
+	output string,
+) {
+	inCh := p.register(input)
+	outCh := p.register(output)
+
+	p.handlers = append(p.handlers, func(ctx context.Context) error {
+		return callback(ctx, inCh, outCh)
+	})
+}
+
+func (p *Pipeline) RegisterMultiplexer(
+	callback func(
+		ctx context.Context,
+		inputs []chan string,
+		output chan string,
+	) error,
+	inputs []string,
+	output string,
+) {
+	if len(inputs) == 0 {
+		panic("empty inputs")
+	}
+
+	inChans := make([]chan string, len(inputs))
+	for i, name := range inputs {
+		inChans[i] = p.register(name)
+	}
+
+	p.handlers = append(p.handlers, func(ctx context.Context) error {
+		return callback(ctx, inChans, p.register(output))
+	})
+}
+
+func (p *Pipeline) RegisterSeparator(
+	callback func(
+		ctx context.Context,
+		input chan string,
+		outputs []chan string,
+	) error,
+	input string,
+	outputs []string,
+) {
+	if len(outputs) == 0 {
+		panic("empty outputs")
+	}
+
+	outChans := make([]chan string, len(outputs))
+	for i, name := range outputs {
+		outChans[i] = p.register(name)
+	}
+
+	inCh := p.register(input)
+
+	p.handlers = append(p.handlers, func(ctx context.Context) error {
+		return callback(ctx, inCh, outChans)
+	})
+}
+
+func (p *Pipeline) Run(ctx context.Context) error {
+	defer func() {
+		for _, ch := range p.channels {
+			close(ch)
+		}
+	}()
+
+	errgr, ctx := errgroup.WithContext(ctx)
+
+	for _, handler := range p.handlers {
+		errgr.Go(func() error {
+			return handler(ctx)
+		})
+	}
+
+	return errgr.Wait() //nolint:wrapcheck
+}
+
+func (p *Pipeline) Send(input string, data string) error {
+	ch, ok := p.channels[input]
+	if !ok {
+		return ErrChanNotFound
+	}
+
+	ch <- data
+
+	return nil
+}
+
+func (p *Pipeline) Recv(output string) (string, error) {
+	channel, okey := p.channels[output]
```

Comment:
```suggestion
	channel, ok := p.channels[output]
```

это общепринято, давай не будем выдумывать. если что - можно использовать `nolint`